blaze-client/src/main/scala/org/http4s/blaze/client/Http1Connection.scala

```diff
@@ -310,105 +310,113 @@ private final class Http1Connection[F[_]](
+      val (rawBody, _): (EntityBody[F], () => Future[ByteBuffer]) =
+        collectBodyFromParser(buffer, terminationConditionParsingPrelude _)
+
+      // to collect the trailers we need a cleanup helper and an effect in the attribute map
+      val (trailerCleanup, attributes): (() => Unit, Vault) =
+        if (parser.getHttpVersion().minor == 1 && parser.isChunked()) {
+          val trailers = new AtomicReference(Headers.empty)
+
+          val attrs = Vault.empty.insert[F[Headers]](
+            Message.Keys.TrailerHeaders[F],
+            F.defer {
+              if (parser.contentComplete()) F.pure(trailers.get())
+              else
+                F.raiseError(
+                  new IllegalStateException(
+                    "Attempted to collect trailers before the body was complete."
                   )
-              },
-            )
-
-            (() => trailers.set(parser.getHeaders()), attrs)
-          } else
-            (
-              { () =>
-                ()
-              },
-              Vault.empty,
-            )
-        }
+                )
+            },
+          )
 
-        if (parser.contentComplete()) {
-          trailerCleanup()
-          cleanup()
-          attributes -> rawBody
+          (() => trailers.set(parser.getHeaders()), attrs)
         } else
-          attributes -> rawBody.onFinalizeCaseWeak {
-            case ExitCase.Completed =>
-              Async.shift(executionContext) *> F.delay { trailerCleanup(); cleanup(); }
-            case ExitCase.Error(_) | ExitCase.Canceled =>
-              Async.shift(executionContext) *> F.delay {
-                trailerCleanup(); cleanup(); stageShutdown()
+          (() => (), Vault.empty)
+
+      if (parser.contentComplete()) {
+        trailerCleanup()
+        cleanupParsingPrelude(closeOnFinish, headers)
+        attributes -> rawBody
+      } else
+        attributes -> rawBody.onFinalizeCaseWeak {
+          case ExitCase.Completed =>
+            Async.shift(executionContext) *>
+              F.delay { trailerCleanup(); cleanupParsingPrelude(closeOnFinish, headers); }
+          case ExitCase.Error(_) | ExitCase.Canceled =>
+            Async.shift(executionContext) *>
+              F.delay {
+                trailerCleanup(); cleanupParsingPrelude(closeOnFinish, headers); stageShutdown()
               }
-          }
-      }
-      cb(
-        Right(
-          Response[F](
-            status = status,
-            httpVersion = httpVersion,
-            headers = headers,
-            body = body.interruptWhen(idleTimeoutS),
-            attributes = attributes,
-          )
+        }
+    }
+
+    cb(
+      Right(
+        Response[F](
+          status = status,
+          httpVersion = httpVersion,
+          headers = headers,
+          body = body.interruptWhen(idleTimeoutS),
+          attributes = attributes,
         )
       )
-    } catch {
+    )
+  }
+
+  private def parsePrelude(
```

Comment:
IMO this class would be easier if reading it top to bottom would be consistent with the order in which these operations happen (methods are called). The already existing methods are ordered like this. At least the most important ones.

I would like to see them in this order:
- runRequest
- executeRequest
- receiveResponse
- handleRead
- parsePrelude
- parsePreludeFinished (new)
- terminationConditionParsingPrelude (new)
- cleanupParsingPrelude (new)